pytensor/graph/basic.py

```diff
@@ -266,8 +266,13 @@ def clone_with_new_inputs(
         assert isinstance(inputs, (list, tuple))
         remake_node = False
         new_inputs: List["Variable"] = list(inputs)
+
+        # Some Ops like Alloc require the node to always be rebuilt in non-strict mode
+        # as the output type depends on the input values and not just their types
+        output_type_depends_on_input_value = self.op._output_type_depends_on_input_value
+
         for i, (curr, new) in enumerate(zip(self.inputs, new_inputs)):
-            if curr.type != new.type:
+            if (curr.type != new.type) or output_type_depends_on_input_value:
```

Comment:
Not super clear the branching logic from the outsider glance, can you please add comments that explain what exactly happens in each branch?

Reply:
Done